common/network-types/src/frame.rs

```diff
@@ -0,0 +1,534 @@
+use crate::errors::NetworkTypeError;
+use crate::errors::NetworkTypeError::{IncompleteFrame, InvalidFrameId, InvalidSegment, ReassemblerClosed};
+use crossbeam_skiplist::SkipSet;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use futures::{Sink, Stream};
+use hopr_platform::time::native::current_time;
+use hopr_primitive_types::prelude::AsUnixTimestamp;
+use serde::{Deserialize, Serialize};
+use std::ops::Sub;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU16, AtomicU64, Ordering};
+use std::sync::OnceLock;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+fn segment(data: &[u8], mtu: u16, frame_id: u16) -> Vec<Segment> {
+    let chunks = data.chunks(mtu as usize);
+    assert!(chunks.len() < u16::MAX as usize, "data too long");
+
+    let seq_len = chunks.len() as u16;
+
+    chunks
+        .enumerate()
+        .map(|(idx, data)| Segment {
+            frame_id,
+            seq_len,
+            seq_idx: idx as u16,
+            data,
+        })
+        .collect()
+}
+
+/// Data frame of arbitrary length.
+/// The frame can be segmented into [segments](Segment) and reassembled back
+/// via [FrameReassembler].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Frame {
+    /// Identifier of this frame.
+    pub frame_id: u16,
+    /// Frame data.
+    pub data: Box<[u8]>,
+}
+
+impl Frame {
+    /// Segments this frame into a list of [segments](Segment) each of maximum sizes `mtu`.
+    #[inline]
+    pub fn segment(&self, mtu: u16) -> Vec<Segment> {
+        segment(self.data.as_ref(), mtu, self.frame_id)
+    }
+}
+
+/// Represents a frame segment.
+/// Besides the data, a segment carries information about the total number of
+/// segments in the original frame, its index within the frame and
+/// ID of that frame.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Segment<'a> {
+    /// Id of the [Frame] this segment belongs to.
+    pub frame_id: u16,
+    /// Index of this segment within the segment sequence.
+    pub seq_idx: u16,
+    /// Total number of segments within this segment sequence.
+    pub seq_len: u16,
+    /// Data in this segment.
+    #[serde(with = "serde_bytes")]
+    pub data: &'a [u8],
+}
+
+impl<'a> PartialOrd<Segment<'a>> for Segment<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Segment<'_> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.frame_id.cmp(&other.frame_id) {
+            std::cmp::Ordering::Equal => self.seq_idx.cmp(&other.seq_idx),
+            cmp => cmp,
+        }
+    }
+}
+
+impl From<Segment<'_>> for Box<[u8]> {
+    fn from(value: Segment<'_>) -> Self {
+        let mut ret = Vec::with_capacity(6 + value.data.len());
```

Comment:
This is not dangerous per se, but risks a reallocation, if the constant value is not updated, if one of the underlying members changes the type, it'd be safer to use https://doc.rust-lang.org/std/mem/fn.size_of_val.html